PR: https://github.com/odoo/odoo/pull/217653

From: 65a61cc982aaad2550300a622f432eb7af51faa7
From: thha-odoo
Date: 2026-09-08 11:37:43

Breaking data model changes scores: del:6 + add:0, change matches:
-    postpone_max_days = fields.Integer(export_string_translation=False,
-    last_executed_carryover_date = fields.Date(export_string_translation=False)
-    actual_lastcall = fields.Date(export_string_translation=False)
-    already_accrued = fields.Boolean()
-    yearly_accrued_amount = fields.Float(export_string_translation=False)
-    expiring_carryover_days = fields.Float("The number of carried over days that will expire on carried_over_days_expiration_date")

Total Changes: 978

[REF] hr_holidays: changing the way accruals days are computed

Accrual allocation initialization issue:
*---------------------------------------

For an accrual allocation, due to `_add_lastcall` initialization method, the first carryover date could be skipped. Indeed this method only includes the following accrual "events" in his `nextcall` computation:
* Carryover days expiration date
* Current level next period
* Next level transition

PS: The job of _add_lastcalls was to guess the value of nextcall, lastcall and actual_lastcall after saving the allocation

Accrual allocation `_process_accrual_plans` issue:
*-------------------------------------------------

For an accrual allocation, to compute the number of days when the "accrued_gain_time" is "start", the code is using a bool field (already_accrued) and was trying to use the same logic as when "accrued_gain_time" is "end", but adding another alternative at the end of the loop. This was a bad idea as it has been proven that it leads to complex, hard to maintain, long, and bug prone code.

Here is the (impossible) challenge of this code:
The function `_process_accrual_plans` tries to update the `number_of_days` according to the accrual "events" (carryover, carryover expiring days, level transition, level period transition-daily-monthly-...). But the problem is that sometimes, it has to process multiple accrual events to compute the `number_of_days` of the allocation in the far future (possible when the user wants to see his number_of_days in the future from the time off dashboard). In this case, when the "accrued_gain_time" is "start" and using 'already_accrued', here's how the function works :
* Run through each accrual event until the nextcall is past the target date (same as when "accrued_gain_time" is "end")
* Run the additionnal alternative

This alternative set 'already_accrued' to True so that the next accrual event in the main loop doesn't accrue the days to the allocation.

But this code also has to take care of each day cron run. In that case, here's how it behaves:
* If the allocation nextcall happens before the target date:
	- Run one iteration of the main loop, which probably doesn't add the accrued days as already_accrued is probably True.
	- Run the additionnal alternative

This means that all the logic has to be at the same time in the main accrual event loop, and in the alternative, which is really hard, and led to the spaghetti we have today (which still contains a lot of bugs).

To avoid this issue, this PR separates the logic of each case ("accrued_gain_time" "start" or "end"), and removes 'already_accrued'.

Accrual allocation `_process_accrual_plans` new feature:
*-------------------------------------------------------

To make it easier to use, this method now returns a dict containing the accrual values of the allocations on the `target_date`, and do not directly apply the changes on the allocation anymore. For instance, this avoid having to create 'fake_allocation' using `new(origin=self)`.

However, it adds a little bit of complexity from the code, as some function now have to pay attention whether they should read the value from the values in the dict representing the data of the allocation, of the fields themselves (see class description for more explanations).

Behavior correction:
*-------------------

Before this PR, a day belonged to a level only if it was between the start of this level (not included), and the start on the following level (included). As the start of the level was not included, this could lead to some small issues like the days are only accrued on the second days of the level, even when "accrued_gain_time" of the accrual plan is "start".

Tests:
*-----

As I spend a lot of time making the tests work, I took this opportunity to make a few changes, mostly to make test simpler to understand, or to correct the tests that were doing wrong assertions.

Accrual allocation refactoring:
*------------------------------

* Deleting the property "already_accrued"
* Renaming
	- lastcall -> last_accrual : as described in the definition of lastcall, this property contains the "Date of the last accrual allocation"
	- actual_lastcall -> lastcall : simpler name
	- postpone_max_days -> max_carriedover_duration: clearer name, moreover, the unit of time of this field can be either 'Days' or 'Hours'
	- expiring_carryover_days -> previous_carryover_number_of_days: the field expiring_carryover_days contains the number of days the allocation had on the last carryover, so this names fits best
* Adding a few fields to the Form: when creating an allocation with an accrual plan in a Form, everything was running smoothly, and the number_of_days was computed correctly. However, when saving the allocation, as the lastcall, actual_lastcall and nextcall were not saved (no fields matching them in the Form), they needed to be recomputed. So the method _add_lastcall was called to try to approximate what were the values of each of these fields depending on the fields.Date.today().
* Getting out of the infinite compute dependency of `number_of_days`, `number_of_days_display` and `number_of_hours_display`. Those fields were 3 computes fields depending on each other (and the orm was not made to use it that way) -> `number_of_days` is not computed anymore, and the 2 other fields set it using 'inverse' (more standard way of doing things).

community-217653
upgrade-8081
task-4508812

closes odoo/odoo#217653

Related: odoo/upgrade#8081
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_holidays/models/hr_employee.py
+++ b/addons/hr_holidays/models/hr_employee.py
@@ -533,11 +533,10 @@ class HrEmployee(models.Model):
             return self.browse(ctx.get('default_employee_id'))
         return self.env.user.employee_id
 
-    def _get_consumed_leaves(self, work_entry_types, target_date=False, ignore_future=False):
-        """ This method won't call `_get_future_leaves_on` for the allocations contained by this variable (it will only use the current value of
-            the `number_of_days` of the allocation, alias `number_of_hours_display`)
-
-            `precomputed_allocations`: context variable (recordset) which can be used to pass allocation that are considered to be already computed
+    def _get_consumed_leaves(self, work_entry_types, target_date=False, ignore_future=False, precomputed_allocations={}):
+        """ The unit of the value returned in the dict is defined by the matching allocation.work_entry_type.unit_of_measure
+            :param precomputed_allocations: this method won't call `_get_additionnal_future_leaves_on` for
+            the allocations contained by this variable (they are considered to be already updated for 'target_date')
         """
         employees = self or self._get_contextual_employee()
         leaves_domain = [
@@ -614,21 +613,18 @@ class HrEmployee(models.Model):
                 'to_recheck_leaves': self.env['hr.leave']
             })
         )
-        precomputed_allocations = self.env.context.get('precomputed_allocations')
         for allocation in allocations:
             allocation_data = allocations_leaves_consumed[allocation.employee_id][allocation.work_entry_type_id][allocation]
-            precomputed = False
-            if precomputed_allocations:
-                if allocation.id in precomputed_allocations.ids:
-                    allocation = precomputed_allocations.filtered(lambda alloc: alloc._origin.id == allocation.id)[0]
-                    precomputed = True
             future_leaves = 0
-            if allocation.accrual_plan_id and not precomputed and not ignore_future:
-                future_leaves = allocation._get_future_leaves_on(target_date)
-            max_leaves = allocation.number_of_hours_display\
-                if allocation.work_entry_type_id.unit_of_measure == 'hour'\
-                else allocation.number_of_days_display
-            max_leaves += future_leaves
+            if precomputed_alloc := precomputed_allocations.get(allocation):
+                max_leaves = allocation._convert_from_type_request_unit(
+                    precomputed_alloc['allocated_duration'], allocation.work_entry_type_id.unit_of_measure, precomputed_alloc)
+            else:
+                if allocation.accrual_plan_id and not ignore_future:
+                    future_leaves = allocation._get_additionnal_future_leaves_on(target_date, precomputed_allocations)
+                max_leaves = (allocation.number_of_hours_display
+                    if allocation.work_entry_type_id.unit_of_measure == 'hour'
+                    else allocation.number_of_days) + future_leaves
             allocation_data.update({
                 'max_leaves': max_leaves,
                 'accrual_bonus': future_leaves,
@@ -641,7 +637,7 @@ class HrEmployee(models.Model):
         for employee in employees:
             for work_entry_type in work_entry_types:
                 if not work_entry_type.requires_allocation:
-                    # Ensure that leave types that do not require allocation are
+                    # Ensure that time types that do not require allocation are
                     # still stored in the consumed allocation leaves.
                     # False is the special key used for this type of leave
                     allocations_leaves_consumed[employee][
@@ -756,7 +752,7 @@ class HrEmployee(models.Model):
                     virtual_remaining = 0
                     additional_leaves_duration = 0
                     for allocation in consumed_content:
-                        latest_accrual_bonus += allocation and allocation._get_future_leaves_on(date_to_simulate)
+                        latest_accrual_bonus += allocation and allocation._get_additionnal_future_leaves_on(date_to_simulate, precomputed_allocations)
                         date_accrual_bonus += consumed_content[allocation]['accrual_bonus']
                         virtual_remaining += consumed_content[allocation]['virtual_remaining_leaves']
                     for leave in content['to_recheck_leaves']:

--- a/addons/hr_holidays/models/hr_leave_accrual_plan_level.py
+++ b/addons/hr_holidays/models/hr_leave_accrual_plan_level.py
@@ -4,6 +4,7 @@ from dateutil.relativedelta import relativedelta
 
 from odoo import _, api, fields, models
 from odoo.exceptions import ValidationError, UserError
+from odoo.tools.date_utils import get_timedelta
 
 
 def _get_selection_days(self):
@@ -104,12 +105,14 @@ class HrLeaveAccrualLevel(models.Model):
         export_string_translation=False)
     cap_accrued_time = fields.Boolean(export_string_translation=False,
         help="When the field is checked the balance of an allocation using this accrual plan will never exceed the specified amount.")
+    # Unit defined by `added_value_type`
     maximum_leave = fields.Float(
         digits=(16, 2), compute="_compute_maximum_leave", default=0, readonly=False, store=True,
         help="Choose a cap for this accrual.", export_string_translation=False)
     cap_accrued_time_yearly = fields.Boolean(export_string_translation=False,
         store=True, readonly=False,
         help="When the field is checked the total amount accrued each year will be capped at the specified amount")
+    # Unit defined by `added_value_type`
     maximum_leave_yearly = fields.Float(digits=(16, 2), export_string_translation=False)
     can_be_carryover = fields.Boolean(related='accrual_plan_id.can_be_carryover', readonly=True,
         export_string_translation=False)
@@ -132,7 +135,8 @@ class HrLeaveAccrualLevel(models.Model):
         default='unlimited', required=True,
         help="You can limit the accrued time carried over for the next period."
     )
-    postpone_max_days = fields.Integer(export_string_translation=False,
+    # Unit defined by `added_value_type`
+    max_carriedover_duration = fields.Integer(export_string_translation=False,
         help="Set a maximum of accruals an allocation keeps at the end of the year.")
     can_modify_value_type = fields.Boolean(compute="_compute_can_modify_value_type", default=False,
         export_string_translation=False)
@@ -155,8 +159,8 @@ class HrLeaveAccrualLevel(models.Model):
         'CHECK(added_value > 0)',
         'You must give a rate greater than 0 in accrual plan levels.',
     )
-    _valid_postpone_max_days_value = models.Constraint(
-        "CHECK(action_with_unused_accruals <> 'all' OR carryover_options <> 'limited' OR COALESCE(postpone_max_days, 0) > 0)",
+    _valid_max_carriedover_duration_value = models.Constraint(
+        "CHECK(action_with_unused_accruals <> 'all' OR carryover_options <> 'limited' OR COALESCE(max_carriedover_duration, 0) > 0)",
         'You cannot have a maximum quantity to carryover set to 0.',
     )
     _valid_accrual_validity_value = models.Constraint(
@@ -296,8 +300,7 @@ class HrLeaveAccrualLevel(models.Model):
         return ['hourly']
 
     def _get_next_date(self, last_call):
-        """
-        Returns the next date with the given last call
+        """ Returns the next date the allocation will be accrued with the given last call (last_call not included)
         """
         self.ensure_one()
         if self.frequency in self._get_hourly_frequencies() + ['daily']:
@@ -365,7 +368,7 @@ class HrLeaveAccrualLevel(models.Model):
             date = last_call + relativedelta(day=int(self.first_day))
             if last_call >= date:
                 return date
-            return last_call + relativedelta(day=int(self.first_day), months=-1, days=1)
+            return last_call + relativedelta(day=int(self.first_day), months=-1)
 
         if self.frequency == 'biyearly':
             first_date = last_call + relativedelta(month=int(self.first_month), day=int(self.first_month_day))
@@ -395,3 +398,7 @@ class HrLeaveAccrualLevel(models.Model):
 
     def action_save_new(self):
         return self.accrual_plan_id.action_create_accrual_plan_level()
+
+    def _get_start_date(self, date_from):
+        self.ensure_one()
+        return date_from + get_timedelta(self.start_count, self.start_type)

--- a/addons/hr_holidays/models/hr_leave_allocation.py
+++ b/addons/hr_holidays/models/hr_leave_allocation.py
@@ -2,21 +2,31 @@
 
 # Copyright (c) 2005-2006 Axelor SARL. (http://www.axelor.com)
 from calendar import monthrange
-from datetime import datetime, date, time
+from datetime import datetime, date, time, timedelta
 from dateutil.relativedelta import relativedelta
 from zoneinfo import ZoneInfo
 
 from odoo import api, fields, models, _
 from odoo.tools import format_date
+from odoo.fields import Domain
 from odoo.addons.hr_holidays.models.hr_leave import get_employee_from_context
 from odoo.exceptions import UserError, ValidationError
-from odoo.fields import Domain
 from odoo.tools.float_utils import float_round
-from odoo.tools.date_utils import get_timedelta
 
 
 class HrLeaveAllocation(models.Model):
-    """ Allocation Requests Access specifications: similar to leave requests """
+    """ Allocation Requests Access specifications: similar to leave requests
+
+        Some of the methods of this class are tagged with 'Possibly accrual inconsistent' or 'Accrual inconsistent'.
+        This means:
+        - They mustn't read allocation fields that are not listed in `_get_accrual_to_update_fields`, but should
+          instead use the values of the parameter `allocation_data`
+        - They mustn't write any field of the allocation, but should update the `allocation_data` parameter
+
+        Some of these methods have `allocation_data={}`. In the case `allocation_data` is set to `{}`, those
+        methods can directly use/modify any of the allocation field. In this case, those method will be tagged
+        'Possibly accrual inconsistent', and not 'Accrual inconsistent'.
+    """
     _name = 'hr.leave.allocation'
     _description = "Time Off Allocation"
     _order = "create_date desc"
@@ -100,17 +110,17 @@ class HrLeaveAllocation(models.Model):
     notes = fields.Text('Reasons', readonly=False)
     # duration
     number_of_days = fields.Float(
-        'Number of Days', compute='_compute_number_of_days', store=True, readonly=False, tracking=True, default=1,
+        'Number of Days', readonly=False, tracking=True, default=1,
         help='Duration in days. Reference field to use when necessary.')
     number_of_days_display = fields.Float(
-        'Duration (days)', compute='_compute_number_of_days_display',
+        'Duration (days)', compute='_compute_number_of_days_display', inverse='_inverse_number_of_days_display',
         help="For an Accrual Allocation, this field contains the theorical amount of time given to the employee, due to a previous start date, on the first run of the plan. This can be manually edited.")
     number_of_hours_display = fields.Float(
         'Duration (hours)', default_export_compatible=True, compute='_compute_number_of_hours_display', store=True,
+        inverse='_inverse_number_of_hours_display',
         help="For an Accrual Allocation, this field contains the theorical amount of time given to the employee, due to a previous start date, on the first run of the plan. This can be manually edited.")
     duration_display = fields.Char('Allocated (Days/Hours)', compute='_compute_duration_display',
         help="Field allowing to see the allocation duration in days or hours depending on the type_request_unit")
-    last_executed_carryover_date = fields.Date(export_string_translation=False)
     # details
     approver_id = fields.Many2one(
         'hr.employee', string='First Approval', readonly=True, copy=False,
@@ -128,21 +138,23 @@ class HrLeaveAllocation(models.Model):
         ('day', 'Day'),
     ], compute="_compute_type_request_unit")
     department_id = fields.Many2one('hr.department', compute='_compute_department_id', store=True, string='Department', readonly=False)
-    # accrual configuration
-    lastcall = fields.Date("Date of the last accrual allocation", readonly=True)
-    # lastcall is only updated on accrual date. On other dates such as carryover date,
-    # actual_lastcall will store the date of the lastcall of the accrual allocation
-    actual_lastcall = fields.Date(export_string_translation=False)
-    nextcall = fields.Date("Date of the next accrual allocation", readonly=True, default=False)
-    already_accrued = fields.Boolean()
-    yearly_accrued_amount = fields.Float(export_string_translation=False)
     is_officer = fields.Boolean(compute='_compute_is_officer')
-    accrual_plan_id = fields.Many2one('hr.leave.accrual.plan', index='btree_not_null', tracking=True)
+    # max_leaves, leaves_taken and virtual_remaining_leaves are for display purpose only
+    # Do not use them for computation, use the result of _process_accrual_plans instead !
     max_leaves = fields.Float(compute='_compute_leaves')
     leaves_taken = fields.Float(compute='_compute_leaves', string='Time off Taken')
     virtual_remaining_leaves = fields.Float(compute='_compute_leaves', string='Available Time Off')
-    expiring_carryover_days = fields.Float("The number of carried over days that will expire on carried_over_days_expiration_date")
-    carried_over_days_expiration_date = fields.Date("Carried over days expiration date")
+
+    # ============================== Accrual type allocation fields ==============================
+    accrual_plan_id = fields.Many2one('hr.leave.accrual.plan', index='btree_not_null', tracking=True)
+    # `last_accrual`, `lastcall` and `nextcall` will be assigned a value when:
+    # `date_from <= first_level_start <= today` AND `_process_accrual_plans` has been called
+    last_accrual = fields.Date("Date of the last accrual allocation", export_string_translation=False)
+    lastcall = fields.Date('Date of the last executed accrual event (accrual, carryover date, expiring carriedover days, ...)', export_string_translation=False)
+    nextcall = fields.Date('Date of the closest next accrual event (accrual, carryover date, expiring carriedover days, ...)', export_string_translation=False)
+    yearly_accrued_days = fields.Float('Number of accrued days since the last carryover', export_string_translation=False)
+    previous_carryover_number_of_days = fields.Float('The number of days on previous carryover date', export_string_translation=False)
+    carried_over_days_expiration_date = fields.Date('Carried over days expiration date', export_string_translation=False)
 
     @api.constrains('date_from', 'date_to')
     def _check_date_from_date_to(self):
@@ -156,6 +168,13 @@ class HrLeaveAllocation(models.Model):
     def _compute_is_officer(self):
         self.is_officer = self.env.user.has_group("hr_holidays.group_hr_holidays_user")
 
+    def _get_employee_hours_per_day(self, allocation_data={}):
+        """ Possibly accrual inconsistent (see the comment under the class declaration) """
+        self.ensure_one()
+        date_from = allocation_data.get('nextcall', self.nextcall) or self.date_from \
+            if self.accrual_plan_id else self.date_from
+        return self.employee_id._get_hours_per_day(date_from)
+
     def _get_title(self):
         self.ensure_one()
         if not self.work_entry_type_id:
@@ -212,7 +231,7 @@ class HrLeaveAllocation(models.Model):
                 )
             allocation.name_validity = name_validity
 
-    @api.depends('employee_id', 'work_entry_type_id')
+    @api.depends('employee_id', 'work_entry_type_id', 'nextcall')
     def _compute_leaves(self):
         date_from = fields.Date.context_today(self)
         employee_days_per_allocation = self.employee_id._get_consumed_leaves(self.work_entry_type_id, date_from, ignore_future=True)[0]
@@ -233,7 +252,7 @@ class HrLeaveAllocation(models.Model):
         for allocation in self:
             if not allocation.employee_id:
                 continue
-            allocation.number_of_hours_display = (allocation.number_of_days * allocation.employee_id._get_hours_per_day(allocation.date_from))
+            allocation.number_of_hours_display = (allocation.number_of_days * allocation._get_employee_hours_per_day())
 
     @api.depends('number_of_hours_display', 'number_of_days_display')
     def _compute_duration_display(self):
@@ -289,14 +308,14 @@ class HrLeaveAllocation(models.Model):
                     default_work_entry_type_id = self._default_work_entry_type_id()
                 allocation.work_entry_type_id = default_work_entry_type_id
 
-    @api.depends('work_entry_type_id', 'number_of_hours_display', 'number_of_days_display', 'type_request_unit', 'employee_id')
-    def _compute_number_of_days(self):
+    def _inverse_number_of_days_display(self):
         for allocation in self:
-            allocation_unit = allocation.type_request_unit
-            if allocation_unit != 'hour':
-                allocation.number_of_days = allocation.number_of_days_display
-            elif allocation_unit == 'hour' and allocation.employee_id:
-                allocation.number_of_days = allocation.number_of_hours_display / allocation.employee_id._get_hours_per_day(allocation.date_from)
+            allocation.number_of_days = allocation.number_of_days_display
+
+    @api.onchange('employee_id')
+    def _inverse_number_of_hours_display(self):
+        for allocation in self.filtered('employee_id'):
+            allocation.number_of_days = allocation.number_of_hours_display / allocation._get_employee_hours_per_day()
 
     @api.constrains('number_of_days', 'work_entry_type_id', 'employee_id')
     def _check_negative_allocation(self):
@@ -343,7 +362,8 @@ class HrLeaveAllocation(models.Model):
         for allocation in self:
             allocation.type_request_unit = allocation._get_request_unit()
 
-    def _get_carryover_date(self, date_from):
+    def _get_next_carryover_date(self, date_from, date_from_included=True):
+        """ Returns the next carry-over date, `date_from` included or not """
         self.ensure_one()
         carryover_time = self.accrual_plan_id.carryover_date
         accrual_plan = self.accrual_plan_id
@@ -357,399 +377,430 @@ class HrLeaveAllocation(models.Model):
             # 2020/2/31 will be changed to 2020/2/29
             day = min(monthrange(date_from.year, month)[1], int(accrual_plan.carryover_day))
             carryover_date = date(date_from.year, month, day)
-        if date_from > carryover_date:
+
+        if date_from > carryover_date or (not date_from_included and carryover_date == date_from):
             carryover_date += relativedelta(years=1)
         return carryover_date
 
-    def _add_days_to_allocation(self, current_level, current_level_maximum_leave, leaves_taken, period_start, period_end):
-        start_date = self.env.context.get('start_date') or self.lastcall
-        end_date = self.env.context.get('end_date') or self.nextcall
-        days_to_add = self._process_accrual_plan_level(
-            current_level, period_start, start_date, period_end, end_date)
+    def _convert_to_type_request_unit(self, duration, duration_unit, allocation_data={}):
+        """ Possibly accrual inconsistent (see the comment under the class declaration) """
+        self.ensure_one()
+        return self._convert_duration(duration, duration_unit, self.type_request_unit, allocation_data)
+
+    def _convert_from_type_request_unit(self, duration, duration_unit, allocation_data={}):
+        """ Possibly accrual inconsistent (see the comment under the class declaration) """
+        self.ensure_one()
+        return self._convert_duration(duration, self.type_request_unit, duration_unit, allocation_data)
+
+    def _convert_duration(self, duration, src_unit, dst_unit, allocation_data={}):
+        """ Possibly accrual inconsistent (see the comment under the class declaration) """
+        self.ensure_one()
+        assert all(unit in ('hour', 'half_day', 'day') for unit in (src_unit, dst_unit))
+        if dst_unit == 'hour' and src_unit in ('day', 'half_day'):
+            return duration * self._get_employee_hours_per_day(allocation_data)
+        if dst_unit in ('day', 'half_day') and src_unit == 'hour':
+            return duration / self._get_employee_hours_per_day(allocation_data)
+        return duration
+
+    def _add_days_to_allocation(self, allocation_data, current_level, period_start, period_end, start_date, end_date):
+        """ Accrual inconsistent (see the comment under the class declaration) """
+        self.ensure_one()
+        added_duration = self._get_period_added_duration(
+            allocation_data, current_level, period_start, start_date, period_end, end_date)
         if current_level.cap_accrued_time_yearly:
-            maximum_leave_yearly = current_level.maximum_leave_yearly\
-                if current_level.added_value_type != 'hour'\
-                else current_level.maximum_leave_yearly / self.employee_id._get_hours_per_day(self.nextcall)
-            yearly_remaining_amount = maximum_leave_yearly - self.yearly_accrued_amount
-            days_to_add = min(days_to_add, yearly_remaining_amount)
+            maximum_leave_yearly = self._convert_to_type_request_unit(
+                current_level.maximum_leave_yearly, current_level.added_value_type, allocation_data)
+            yearly_remaining_duration = max(0, maximum_leave_yearly - allocation_data['yearly_accrued_duration'])
+            added_duration = min(added_duration, yearly_remaining_duration)
         if current_level.cap_accrued_time:
-            capped_total_balance = leaves_taken + current_level_maximum_leave
-            days_to_add = min(days_to_add, capped_total_balance - self.number_of_days)
-        self.number_of_days += days_to_add
-        self.yearly_accrued_amount += days_to_add
-
-    def _get_current_accrual_plan_level_id(self, date, level_ids=False):
+            lvl_max_leave_duration = self._convert_to_type_request_unit(
+                current_level.maximum_leave, current_level.added_value_type, allocation_data)
+            capped_total_balance = allocation_data['leaves_taken'] + lvl_max_leave_duration
+            added_duration = min(added_duration, capped_total_balance - allocation_data['allocated_duration'])
+        allocation_data['allocated_duration'] += added_duration
+        allocation_data['yearly_accrued_duration'] += added_duration
+        allocation_data['last_accrual'] = allocation_data['nextcall']
+
+    def _get_current_accrual_plan_levels(self, target_date) -> dict[str, tuple] | None:
+        """ Returns a dict of tuples (lvl, idx) containing the level(s) we are currently in depending on `target_date` parameter.
+            It can return up to 2 entries (when target_date is a level transition), "current_level" and "previous_level" (optional).
+            Return example:  {
+                "previous_level": (lvl3, lvl3_idx),         -> "previous_level" is an optional entry
+                "current_level": (lvl4, lvl4_idx),
+            }
         """
-        Returns a pair (accrual_plan_level, idx) where accrual_plan_level is the level for the given date
-        and idx is the index for the plan in the ordered set of levels
+        self.ensure_one()
+        if not self.accrual_plan_id or not self.accrual_plan_id.level_ids or (self.date_to and target_date > self.date_to):
+            return None
+        lvls_boundaries = self._get_lvls_boundaries()
+        if target_date < lvls_boundaries[0]:
+            return None
+
+        sorted_levels = self.accrual_plan_id.level_ids
+        current_lvl_start = False
+        current_lvl_idx = False
+        for lvl_idx, lvl_start in enumerate(lvls_boundaries):
+            if target_date >= lvl_start:
+                current_lvl_idx = lvl_idx
+                current_lvl_start = lvl_start
+
+        # if `target_date` is a level transition
+        if current_lvl_idx > 0 and target_date == current_lvl_start:
+            return {
+                "previous_level": (sorted_levels[current_lvl_idx - 1], current_lvl_idx - 1),
+                "current_level": (sorted_levels[current_lvl_idx], current_lvl_idx),
+            }
+
+        return {"current_level": (sorted_levels[current_lvl_idx], current_lvl_idx)}
+
+    def _get_current_accrual_plan_level_idx(self, date):
+        """ Returns (current_level, index). If `date` is a levels transition, the returned `current_level` is the
+            one that starts later of the two levels """
+        self.ensure_one()
+        current_levels = self._get_current_accrual_plan_level_idx_for_accrual(date)
+        return current_levels['current_level'] if current_levels else (None, None)
+
+    def _get_current_accrual_plan_level_idx_for_accrual(self, target_date):
+        """ Unlike `_get_current_accrual_plan_level_idx`, takes the levels transition into account.
+            For example, if `date` is a level transition and `accrued_gain_time` is "end", then we want to
+            get the "previous" plan level as we need to compute `number_of_days`, `expiring_days`, ... depending on the
+            "previous" level policy.
+
+            Returns a `dict` with the following entries:
+            - "current_level": (current_level, current_level_idx)     -> Current level
+            - "accrual_level": (accrual_level, accrual_level_idx)     -> Current level considering levels transition
         """
         self.ensure_one()
-        if not self.accrual_plan_id.level_ids:
-            return (False, False)
-        # Sort by sequence which should be equivalent to the level
-        if not level_ids:
-            level_ids = self.accrual_plan_id.level_ids.sorted('sequence')
-        current_level = False
-        current_level_idx = -1
-        for idx, level in enumerate(level_ids):
-            if date > self.date_from + get_timedelta(level.start_count, level.start_type):
-                current_level = level
-                current_level_idx = idx
-        # If transition_mode is set to `immediately` or we are currently on the first level
-        # the current_level is simply the first level in the list.
-        if current_level_idx <= 0 or self.accrual_plan_id.transition_mode == "immediately":
-            return (current_level, current_level_idx)
-        # In this case we have to verify that the 'previous level' is not the current one due to `end_of_accrual`
-        level_start_date = self.date_from + get_timedelta(current_level.start_count, current_level.start_type)
-        previous_level = level_ids[current_level_idx - 1]
-        # If the next date from the current level's start date is before the last call of the previous level
-        # return the previous level
-        if current_level._get_next_date(level_start_date) < previous_level._get_next_date(level_start_date):
-            return (previous_level, current_level_idx - 1)
-        return (current_level, current_level_idx)
+        plan_levels = self._get_current_accrual_plan_levels(target_date)
+        if not plan_levels:
+            return None
+        accrual_level = plan_levels['previous_level'] if 'previous_level' in plan_levels\
+            and self.accrual_plan_id.accrued_gain_time == 'end' else plan_levels['current_level']
+        return {'current_level': plan_levels['current_level'], 'accrual_level': accrual_level}
 
     def _get_accrual_plan_level_work_entry_prorata(self, level, start_period, start_date, end_period, end_date):
+        """ Possibly accrual inconsistent (see the comment under the class declaration) """
         self.ensure_one()
-        datetime_min_time = datetime.min.time()
+        if level.frequency not in level._get_hourly_frequencies() and not level.accrual_plan_id.is_based_on_worked_time:
+            return 1
+
         version = self.employee_id._get_version(start_date)
         resource_tz = ZoneInfo(version._get_tz() or 'UTC')
-        start_dt = datetime.combine(start_date, datetime_min_time, tzinfo=resource_tz)
-        end_dt = datetime.combine(end_date, datetime_min_time, tzinfo=resource_tz)
-        leaves_eligible = self.employee_id.sudo()._get_leave_days_data_batch(start_dt, end_dt,
-            calendar=self.employee_id._get_calendars(start_dt)[self.employee_id.id],
+        start_dt = datetime.combine(start_date, time.min, tzinfo=resource_tz)
+        end_dt = datetime.combine(end_date, time.min, tzinfo=resource_tz)
+        calendar = self.employee_id._get_calendars(start_dt)[self.employee_id.id]
+        worked_hours = self.employee_id._get_work_days_data_batch(start_dt, end_dt, calendar=calendar)[self.employee_id.id]['hours']
+        worked_hours += self.employee_id.sudo()._get_leave_days_data_batch(start_dt, end_dt, calendar=calendar,
             domain=[('count_as', '=', 'absence'), ('elligible_for_accrual_rate', '=', True)])[self.employee_id.id]['hours']
-        worked = self.employee_id._get_work_days_data_batch(start_dt, end_dt,
-            calendar=self.employee_id.resource_calendar_id)[self.employee_id.id]['hours']
-        worked += leaves_eligible
         if start_period != start_date or end_period != end_date:
             version = self.employee_id._get_version(start_period)
             resource_tz = ZoneInfo(version._get_tz() or 'UTC')
-            start_dt = datetime.combine(start_period, datetime_min_time, tzinfo=resource_tz)
-            end_dt = datetime.combine(end_period, datetime_min_time, tzinfo=resource_tz)
-            leaves_eligible = self.employee_id.sudo()._get_leave_days_data_batch(start_dt, end_dt,
-                calendar=self.employee_id._get_calendars(start_dt)[self.employee_id.id],
+            start_dt = datetime.combine(start_period, time.min, tzinfo=resource_tz)
+            end_dt = datetime.combine(end_period, time.min, tzinfo=resource_tz)
+            period_worked_hours = self.employee_id._get_work_days_data_batch(start_dt, end_dt, calendar=calendar)[self.employee_id.id]['hours']
+            period_worked_hours += self.employee_id.sudo()._get_leave_days_data_batch(start_dt, end_dt, calendar=calendar,
                 domain=[('count_as', '=', 'absence'), ('elligible_for_accrual_rate', '=', True)])[self.employee_id.id]['hours']
-            planned_worked = self.employee_id._get_work_days_data_batch(start_dt, end_dt,
-                calendar=self.employee_id.resource_calendar_id)[self.employee_id.id]['hours']
-            planned_worked += leaves_eligible
         else:
-            planned_worked = worked
-        left = self.employee_id.sudo()._get_leave_days_data_batch(start_dt, end_dt,
-            calendar=self.employee_id._get_calendars(start_dt)[self.employee_id.id],
+            period_worked_hours = worked_hours
+
+        left = self.employee_id.sudo()._get_leave_days_data_batch(start_dt, end_dt, calendar=calendar,
             domain=[('count_as', '=', 'absence'), ('elligible_for_accrual_rate', '=', False)])[self.employee_id.id]['hours']
         if level.frequency in level._get_hourly_frequencies():
-            if level.accrual_plan_id.is_based_on_worked_time:
-                work_entry_prorata = planned_worked
-            else:
-                work_entry_prorata = planned_worked + left
-        else:
-            work_entry_prorata = worked / (left + planned_worked) if (left + planned_worked) else 0
-        return work_entry_prorata
+            return period_worked_hours if level.accrual_plan_id.is_based_on_worked_time \
+                else period_worked_hours + left
+        return worked_hours / (left + period_worked_hours) if (left + period_worked_hours) else 0
+
+    def _get_period_added_duration(self, allocation_data, level, period_start, start_date, period_end, end_date):
+        """ Accrual inconsistent (see the comment under the class declaration) """
+        self.ensure_one()
+        work_entry_prorata = self._get_accrual_plan_level_work_entry_prorata(level, period_start, start_date, period_end, end_date)
+        added_value = work_entry_prorata * level.added_value
+        added_duration = self._convert_to_type_request_unit(added_value, level.added_value_type, allocation_data)
+        period_prorata = 1
+        if (period_start != start_date or period_end != end_date) and not level.accrual_plan_id.is_based_on_worked_time:
+            period_days = period_end - period_start + timedelta(days=1)
+            days = end_date - start_date + timedelta(days=1)
+            period_prorata = min(1, days / period_days) if period_days else 1
+        return added_duration * period_prorata
+
+    @api.model
+    def _get_days_after_expiration(self, allocation_data):
+        """ Accrual inconsistent (see the comment under the class declaration) """
+        expiring_days = allocation_data['previous_carryover_allocated_duration'] - allocation_data['leaves_taken']
+        return max(allocation_data['allocated_duration'] - expiring_days, allocation_data['leaves_taken'])
 
-    def _process_accrual_plan_level(self, level, start_period, start_date, end_period, end_date):
+    def _process_carryover_date(self, allocation_data, current_level, carryover_date):
+        """ Accrual inconsistent (see the comment under the class declaration) """
+        self.ensure_one()
+        if current_level.action_with_unused_accruals == 'lost' or not self.accrual_plan_id.can_be_carryover:
+            allocation_data['allocated_duration'] = allocation_data['leaves_taken']
+        else:
+            if current_level.accrual_validity:
+                allocation_data['carried_over_days_expiration_date'] = carryover_date + relativedelta(
+                    **{current_level.accrual_validity_type + 's': current_level.accrual_validity_count})
+
+            if current_level.carryover_options == 'limited':
+                left_allocated_duration = allocation_data['allocated_duration'] - allocation_data['leaves_taken']
+                max_carriedover_duration = self._convert_to_type_request_unit(
+                    current_level.max_carriedover_duration, current_level.added_value_type, allocation_data)
+                carriedover_duration = min(max_carriedover_duration, left_allocated_duration)
+                allocation_data['allocated_duration'] = carriedover_duration + allocation_data['leaves_taken']
+        allocation_data['previous_carryover_allocated_duration'] = allocation_data['allocated_duration']
+
+    def _process_accrual_start(self, allocation_data, current_level, next_accrual):
+        """ Add duration to this allocation if `allocation_data['nextcall']` is an accrual date
+            Called only on accrual allocations with `accrued_gain_time=start`
+            Accrual inconsistent (see the comment under the class declaration)
         """
-        Returns the added days for that level
+        self.ensure_one()
+        period_start = current_level._get_previous_date(allocation_data['nextcall'])
+        current_lvl_start_date = current_level._get_start_date(self.date_from)
+        if allocation_data['nextcall'] == period_start or allocation_data['nextcall'] == current_lvl_start_date:
+            accrual_start_date = allocation_data['nextcall']
+            accrual_end_date = next_accrual
+            period_end = current_level._get_next_date(allocation_data['nextcall'])
+            if current_level.frequency not in current_level._get_hourly_frequencies() + ['daily']:
+                accrual_end_date -= relativedelta(days=1)
+                period_end -= relativedelta(days=1)
+            self._add_days_to_allocation(allocation_data, current_level, period_start, period_end, accrual_start_date, accrual_end_date)
+
+    def _process_accrual_end(self, allocation_data, current_level, current_level_idx, first_level_start_date):
+        """ Add duration to this allocation if `allocation_data['nextcall']` is an accrual date
+            Called only on accrual allocations with `accrued_gain_time=end`
+            Accrual inconsistent (see the comment under the class declaration)
         """
         self.ensure_one()
-        if level.frequency in level._get_hourly_frequencies() or level.accrual_plan_id.is_based_on_worked_time:
-            work_entry_prorata = self._get_accrual_plan_level_work_entry_prorata(level, start_period, start_date, end_period, end_date)
-            added_value = work_entry_prorata * level.added_value
-        else:
-            added_value = level.added_value
-        # Convert time in hours to time in days in case the level is encoded in hours
-        if level.added_value_type == 'hour':
-            added_value = added_value / self.employee_id._get_hours_per_day(self.date_from)
-        period_prorata = 1
-        if (start_period != start_date or end_period != end_date) and not level.accrual_plan_id.is_based_on_worked_time:
-            period_days = (end_period - start_period)
-            call_days = (end_date - start_date)
-            period_prorata = min(1, call_days / period_days) if period_days else 1
-        return added_value * period_prorata
+        next_date = current_level._get_next_date(allocation_data['lastcall'])
+        current_level_last_date = self._get_next_lvl_start(current_level_idx)
+        if allocation_data['nextcall'] == next_date or allocation_data['nextcall'] == current_level_last_date:
+            accrual_start_date = allocation_data['last_accrual'] or first_level_start_date
+            accrual_end_date = allocation_data['nextcall']
+            period_start = current_level._get_previous_date(allocation_data['lastcall'])
+            period_end = next_date
+            if current_level.frequency not in current_level._get_hourly_frequencies() + ['daily']:
+                accrual_end_date -= relativedelta(days=1)
+                period_end -= relativedelta(days=1)
+            self._add_days_to_allocation(allocation_data, current_level, period_start, period_end, accrual_start_date, accrual_end_date)
+
+    def _has_expiring_days(self, allocation_data):
+        """ Accrual inconsistent (see the comment under the class declaration) """
+        self.ensure_one()
+        expiration_date = allocation_data['carried_over_days_expiration_date']
+        return allocation_data['nextcall'] and expiration_date and allocation_data['nextcall'] <= expiration_date
 
-    def _process_accrual_plans(self, date_to=False, force_period=False, log=True):
+    def _init_accrual_calls(self, allocation_data):
+        """ Initialize `last_accrual`, `lastcall` and `nextcall`
+            Accrual inconsistent (see the comment under the class declaration)
         """
-        This method is part of the cron's process.
-        The goal of this method is to retroactively apply accrual plan levels and progress from nextcall to date_to or today.
-        If force_period is set, the accrual will run until date_to in a prorated way (used for end of year accrual actions).
+        self.ensure_one()
+        first_level = self.accrual_plan_id.level_ids[0]
+        lastcall = first_level._get_start_date(self.date_from)
+        if self.accrual_plan_id.accrued_gain_time == 'start':
+            nextcall = lastcall
+        else:
+            nextcall = first_level._get_next_date(lastcall)
+            if self.accrual_plan_id.can_be_carryover:
+                carryover_date = self._get_next_carryover_date(lastcall)
+                nextcall = min(carryover_date, nextcall)
+            if len(self.accrual_plan_id.level_ids) > 1:
+                second_level_start_date = self._get_next_lvl_start(level_idx=0)
+                nextcall = min(second_level_start_date, nextcall)
+        allocation_data['last_accrual'] = None
+        allocation_data['lastcall'] = lastcall
+        allocation_data['nextcall'] = nextcall
+
+    def _get_lvls_boundaries(self) -> list[date] | None:
+        """ Returns a list containing intervals boundaries of each level.
+            The start of the level x is the end date of the previous level.
+
+            For instance, if there are 2 levels, this function will return a list of 2 dates:
+            the start of the first level and the start of the second level
         """
-        def _get_leaves_taken(allocation):
-            precomputed_allocations = allocation
-            if context_precomputed := self.env.context.get('precomputed_allocations'):
-                precomputed_allocations |= context_precomputed
-            # By setting `precomputed_allocations`, avoid infinite loop (otherwise _get_consumed_leaves -> _get_future_leaves_on -> _process_accrual_plans -> ...)
-            employee_days_per_allocation = allocation.employee_id.with_context(precomputed_allocations=precomputed_allocations)._get_consumed_leaves(
-                allocation.work_entry_type_id, allocation.nextcall, ignore_future=True)[0]
-            origin = allocation._origin
-            leaves_taken = employee_days_per_allocation[origin.employee_id][origin.work_entry_type_id][origin]['leaves_taken']
-            return leaves_taken
-
-        def _get_period_start_date(allocation, period_start, current_level, current_level_idx, level_ids, first_level_start_date):
-            start_date = period_start
-            if current_level_idx > 0:
-                prev_level_end = current_level._get_level_transition_date(allocation.date_from)
-                if allocation.accrual_plan_id.transition_mode == 'end_of_accrual':
-                    prev_level = level_ids[current_level_idx - 1]
-                    # If period doesn't end on level transition then get next period end
-                    if prev_level._get_previous_date(prev_level_end) != prev_level_end:
-                        prev_level_end = prev_level._get_next_date(prev_level_end)
-                start_date = max(start_date, prev_level_end)
-            else:
-                start_date = max(start_date, first_level_start_date)
+        self.ensure_one()
+        if self.accrual_plan_id.transition_mode == 'immediately':
+            return self.accrual_plan_id.level_ids.mapped(lambda lvl: lvl._get_start_date(self.date_from))
 
-        date_to = date_to or fields.Date.context_today(self)
-        already_accrued = {allocation.id: allocation.already_accrued or (allocation.number_of_days != 0 and allocation.accrual_plan_id.accrued_gain_time == 'start') for allocation in self}
-        first_allocation = _("""This allocation have already ran once, any modification won't be effective to the days allocated to the employee. If you need to change the configuration of the allocation, delete and create a new one.""")
-        for allocation in self:
-            expiration_date = False
-            if not allocation.accrual_plan_id:
-                continue
+        boundaries = [self.accrual_plan_id.level_ids[0]._get_start_date(self.date_from)]
+        for level in self.accrual_plan_id.level_ids[1:]:
+            expected_lvl_start = level._get_start_date(self.date_from)
+            lvl_start = level._get_next_date(expected_lvl_start + relativedelta(days=-1))
+            boundaries.append(lvl_start)
+        return boundaries
+
+    def _get_next_lvl_start(self, level_idx):
+        self.ensure_one()
+        lvls_boundaries = self._get_lvls_boundaries()
+        if level_idx == len(lvls_boundaries) - 1:
+            return None
+        return lvls_boundaries[level_idx + 1]
+
+    def _process_accrual_plans(self, date_to=None, precomputed_allocations={}):
+        """ This method is part of the cron's process.
+            The goal of this method is to retroactively apply accrual plan levels and progress from nextcall to `date_to` or `today`.
+        """
+        def _get_leaves_taken(allocation, allocation_data):
+            if len(precomputed_allocations):
+                allocations = dict(precomputed_allocations)
+                allocations[allocation] = allocation_data
+            else:
+                allocations = {allocation: allocation_data}
+            # By setting `precomputed_allocations`, avoid infinite loop (otherwise _get_consumed_leaves -> _get_additionnal_future_leaves_on -> _process_accrual_plans -> ...)
+            employee_days_per_allocation = allocation.employee_id._get_consumed_leaves(
+                allocation.work_entry_type_id, allocation_data['nextcall'], ignore_future=True,
+                precomputed_allocations=allocations)[0]
+            leaves_taken = employee_days_per_allocation[allocation.employee_id][allocation.work_entry_type_id][allocation]['leaves_taken']
+            return allocation._convert_to_type_request_unit(
+                leaves_taken, allocation.work_entry_type_id.unit_of_measure or 'day', allocation_data)
+
+        accrual_allocations = self.filtered('accrual_plan_id')
+        precomputed_alloc_recordset = accrual_allocations.filtered(lambda alloc: alloc in precomputed_allocations)
+        allocations_data = (accrual_allocations - precomputed_alloc_recordset)._get_accrual_allocation_data()
+        allocations_data.update({allocation: dict(precomputed_allocations[allocation]) for allocation in precomputed_allocations})
+        accrual_plan_with_levels = accrual_allocations.filtered('accrual_plan_id.level_ids')
+        for allocation in accrual_plan_with_levels:
+            allocation_data = allocations_data[allocation]
             level_ids = allocation.accrual_plan_id.level_ids
-            if not level_ids:
-                continue
-            # "cache" leaves taken, as it gets recomputed every time allocation.number_of_days is assigned to. Without this,
-            # every loop will take 1+ second. It can be removed if computes don't chain in a way to always reassign accrual plan
-            # even if the value doesn't change. This is the best performance atm.
             first_level = level_ids[0]
-            first_level_start_date = allocation.date_from + get_timedelta(first_level.start_count, first_level.start_type)
-            allocation.already_accrued = already_accrued[allocation.id]
-            # first time the plan is run, initialize nextcall and take carryover / level transition into account
-            if not allocation.nextcall:
-                # Accrual plan is not configured properly or has not started
-                if date_to < first_level_start_date:
-                    continue
-                allocation.lastcall = max(allocation.lastcall, first_level_start_date)
-                allocation.actual_lastcall = allocation.lastcall
-                allocation.nextcall = first_level._get_next_date(allocation.lastcall)
-                # adjust nextcall for carryover
-                carryover_date = allocation._get_carryover_date(allocation.nextcall)
-                allocation.nextcall = min(carryover_date, allocation.nextcall)
-                # adjust nextcall for level_transition
-                if len(level_ids) > 1:
-                    second_level_start_date = allocation.date_from + get_timedelta(level_ids[1].start_count, level_ids[1].start_type)
-                    allocation.nextcall = min(second_level_start_date, allocation.nextcall)
-                if log:
-                    allocation._message_log(body=first_allocation)
-            (current_level, current_level_idx) = (False, 0)
-            current_level_maximum_leave = 0.0
-            # all subsequent runs, at every loop:
-            # get current level and normal period boundaries, then set nextcall, adjusted for level transition and carryover
-            # add days, trimmed if there is a maximum_leave
-            while allocation.nextcall <= date_to:
-                if allocation.work_entry_type_id.unit_of_measure == 'day':
-                    leaves_taken = _get_leaves_taken(allocation)
-                else:
-                    leaves_taken = _get_leaves_taken(allocation) / allocation.employee_id._get_hours_per_day(allocation.nextcall or allocation.date_from)
-                (current_level, current_level_idx) = allocation._get_current_accrual_plan_level_id(allocation.nextcall)
-                if not current_level:
-                    break
-                if current_level.cap_accrued_time:
-                    if current_level.added_value_type == "day":
-                        current_level_maximum_leave = current_level.maximum_leave
-                    else:
-                        current_level_maximum_leave = current_level.maximum_leave / allocation.employee_id._get_hours_per_day(allocation.nextcall or allocation.date_from)
-                nextcall = current_level._get_next_date(allocation.nextcall)
-                period_start = current_level._get_previous_date(allocation.lastcall)
-                period_end = current_level._get_next_date(allocation.lastcall)
-                # There are 3 cases where nextcall could be closer than the normal period:
-                # 1. Passing from one level to another, if mode is set to 'immediately'
-                current_level_last_date = False
-                on_level_transition = False
-                if current_level_idx < (len(level_ids) - 1) and allocation.accrual_plan_id.transition_mode == 'immediately':
-                    next_level = level_ids[current_level_idx + 1]
-                    current_level_last_date = allocation.date_from + get_timedelta(next_level.start_count, next_level.start_type)
-                    if allocation.nextcall != current_level_last_date:
-                        nextcall = min(nextcall, current_level_last_date)
-                    else:
-                        on_level_transition = True
-                # 2. On carry-over date
-                carryover_date = allocation._get_carryover_date(allocation.nextcall)
-                if allocation.nextcall < carryover_date < nextcall:
+            first_level_start_date = first_level._get_start_date(allocation.date_from)
+            date_to = date_to or date.today()
+            if date_to < first_level_start_date:
+                allocation_data['leaves_taken'] = 0
+                continue
+
+            if not allocation_data['nextcall']:
+                # First time the plan is run for this allocation
+                allocation._init_accrual_calls(allocation_data)
+
+            while allocation_data['nextcall'] <= date_to:
+                allocation_data['leaves_taken'] = _get_leaves_taken(allocation, allocation_data)
+                plan_levels = allocation._get_current_accrual_plan_level_idx_for_accrual(allocation_data['nextcall'])
+                (current_level, current_level_idx), (accrual_level, accrual_level_idx) = plan_levels["current_level"], plan_levels["accrual_level"]
+                nextcall = current_level._get_next_date(allocation_data['nextcall'])
+
+                # There are 3 more accrual "events" (added to the start-end of each period of each level):
+                # Level transition, carryover date and carriedover expiring days date
+                # 1. Take level transition into account
+                current_level_last_date = allocation._get_next_lvl_start(current_level_idx)
+                if current_level_last_date and allocation_data['nextcall'] < current_level_last_date:
+                    nextcall = min(nextcall, current_level_last_date)
+
+                next_accrual = nextcall
+                carryover_date = allocation._get_next_carryover_date(allocation_data['nextcall'])
+                # 2. Take carryover date into account
+                if allocation_data['nextcall'] == carryover_date and allocation_data['nextcall'] != allocation.date_from \
+                        and allocation_data['nextcall'] != first_level_start_date:
+                    allocation._process_carryover_date(allocation_data, current_level, carryover_date)
+                elif allocation_data['nextcall'] != carryover_date:
                     nextcall = min(nextcall, carryover_date)
 
-                if current_level.accrual_validity:
-                    # 3. On carried over days expiration date
-                    expiration_date = allocation.carried_over_days_expiration_date
-                    # - not expiration_date -> expiration_date needs to be initialized.
-                    # - allocation.nextcall > expiration_date -> the expiration date has passed and the new one should be computed.
-                    # - allocation.expiring_carryover_days == 0 -> If the carryover date of the accrual plan was changed or if a level
-                    #   transition occurred, then the expiration date needs to be updated. However, if allocation.expiring_carryover_days != 0,
-                    #   then this means that some days will expire on expiration_date and that expiration date should be respected and
-                    #   Expiration date will be updated correctly when allocation.nextcall is greater than expiration_date.
-                    if not expiration_date or allocation.nextcall > expiration_date or allocation.expiring_carryover_days == 0:
-                        expiration_date = carryover_date + relativedelta(**{current_level.accrual_validity_type + 's': current_level.accrual_validity_count})
-                        allocation.carried_over_days_expiration_date = expiration_date
-                    if allocation.nextcall < expiration_date < nextcall:
-                        nextcall = expiration_date
-                    if allocation.nextcall == expiration_date:
-                        # Given that allocation.number_of_days = employee time off balance + leaves_taken. So,
-                        # the leaves_taken are included in allocation.number_of_days.
-                        # Also, allocation.expiring_carryover_days includes the leaves_taken before the carryover date
-                        # and allocation.leaves_taken includes all the leaves_taken before the carryover date + all the leaves_taken
-                        # between the carryover date and the expiration_date. So, the number of expiring days will be
-                        # allocation.expiring_carryover_days - allocation.leaves_taken or 0 if all the expiring days were used
-                        # to take time off.
-                        # This ensures that only the days that weren't used to take time off will expire.
-                        expiring_days = max(0, allocation.expiring_carryover_days - leaves_taken)
-                        allocation.number_of_days = max(0, allocation.number_of_days - expiring_days)
-                        allocation.expiring_carryover_days = 0
-
-                is_accrual_date = allocation.nextcall == period_end or allocation.nextcall == current_level_last_date
-                if not allocation.already_accrued and (is_accrual_date or on_level_transition) and \
-                        allocation.accrual_plan_id.accrued_gain_time == 'start':
-                    start_date = _get_period_start_date(allocation, period_start, current_level, current_level_idx, level_ids, first_level_start_date)
-                    allocation.with_context({'start_date': start_date})._add_days_to_allocation(
-                        current_level, current_level_maximum_leave, leaves_taken, period_start, period_end)
-
-                if allocation.nextcall == carryover_date:
-                    allocation.last_executed_carryover_date = carryover_date
-                    if current_level.action_with_unused_accruals == 'lost' or current_level.carryover_options == 'limited':
-                        allocated_days_left = allocation.number_of_days - leaves_taken
-                        allocation_max_days = 0 # default if unused_accrual are lost
-                        if current_level.carryover_options == 'limited':
-                            if current_level.added_value_type == 'day':
-                                postpone_max_days = current_level.postpone_max_days
-                            else:
-                                postpone_max_days = current_level.postpone_max_days / allocation.employee_id._get_hours_per_day(allocation.date_from)
-                            allocation_max_days = min(postpone_max_days, allocated_days_left)
-                        allocation.number_of_days = min(allocation.number_of_days, allocation_max_days) + leaves_taken
-                    allocation.expiring_carryover_days = allocation.number_of_days
-
-                if not allocation.already_accrued and (is_accrual_date or on_level_transition)\
-                        and allocation.accrual_plan_id.accrued_gain_time == 'end':
-                    start_date = _get_period_start_date(allocation, period_start, current_level, current_level_idx, level_ids, first_level_start_date)
-                    allocation.with_context({'start_date': start_date})._add_days_to_allocation(
-                        current_level, current_level_maximum_leave, leaves_taken, period_start, period_end)
-
-                if allocation.nextcall == carryover_date:
-                    allocation.yearly_accrued_amount = 0
-
-                # 1. When accrued_gain_time == 'start', all the days are accrued on the start of the accrual period. For example, if the accrual period
-                #    is from 01/01/2023 to 01/01/2024, then the days will be accrued on 01/01/2023. Given that the carryover date will be >= the start of the accrual period
-                #    (01/01/2023 in the example) the carryover policy should apply to any day accrued during the period from 01/01/2023 to 01/01/2024.
-                # 2.However, if a level transistion occurred, the carryover policy should apply to the days that were accrued during the carryover level only.
-                #   Any days accrued after the carryover level should be excluded.
-                #   So, if carryover date was 01/06/2023, it should be applied to any day accrued between 01/01/2023 and 01/01/2024. If a level transition
-                #   occurred on 01/09/2023 for example, then the carryover should be applied to any day accrued between 01/01/2023 and 01/09/2023.
-                # 3. The following if block will handle the carryover for days accrued after carryover_date until carryover_period_end. Carryover period end is
-                #    adjusted if a level transition occurred. The carryover for days accrued before carryover_date is handled above.
-                if allocation.accrual_plan_id.accrued_gain_time == 'start' and allocation.last_executed_carryover_date:
-                    last_carryover_date = allocation.last_executed_carryover_date
-                    carryover_level, carryover_level_idx = allocation._get_current_accrual_plan_level_id(last_carryover_date)
-                    carryover_period_start = carryover_level._get_previous_date(last_carryover_date)
-                    carryover_period_end = carryover_level._get_next_date(last_carryover_date)
-                    # Adjust carryover_period_end based on level_transition.
-                    if carryover_level_idx < (len(level_ids) - 1) and allocation.accrual_plan_id.transition_mode == 'immediately':
-                        next_level = level_ids[carryover_level_idx + 1]
-                        carryover_level_last_date = allocation.date_from + get_timedelta(next_level.start_count, next_level.start_type)
-                        carryover_period_end = min(carryover_period_end, carryover_level_last_date)
-                    # Handle the special case for hourly/daily accruals. Carryover_period_end should be equal to last_carryover_date
-                    # because the carryover period is just 1 day.
-                    if carryover_level.frequency in carryover_level._get_hourly_frequencies() + ['daily']:
-                        carryover_period_end = last_carryover_date
-                    # Carryover policy should be only applied to the days accrued on period_end.
-                    # Days accrued on level transition date aren't subject to the carryover policy.
-                    # That is why (allocation.nextcall == period_end) is used instead of (is_accrual_date)
-                    accrued = not allocation.already_accrued and allocation.nextcall == period_end
-                    # If the days were accrued on the carryover period, then apply the carryover policy
-                    # If allocation.actual_lastcall == carryover_period_start, it means this loop has already been run once (skip to avoid applying the carryover twice)
-                    if accrued and last_carryover_date <= allocation.nextcall <= carryover_period_end and allocation.actual_lastcall != carryover_period_start:
-                        if carryover_level.action_with_unused_accruals == 'lost' or carryover_level.carryover_options == 'limited':
-                            allocation.last_executed_carryover_date = carryover_date
-                            allocated_days_left = allocation.number_of_days - leaves_taken
-                            postpone_max_days = current_level.postpone_max_days if current_level.added_value_type == 'day' \
-                                else current_level.postpone_max_days / allocation.employee_id._get_hours_per_day(allocation.date_from)
-                            allocated_days_left = allocation.number_of_days - leaves_taken
-                            allocation_max_days = 0 # default if unused_accrual are lost
-                            if current_level.carryover_options == 'limited':
-                                postpone_max_days = current_level.postpone_max_days
-                                allocation_max_days = min(postpone_max_days, allocated_days_left)
-                            allocation.number_of_days = min(allocation.number_of_days, allocation_max_days) + leaves_taken
-
-                if is_accrual_date:
-                    allocation.lastcall = allocation.nextcall
-                allocation.actual_lastcall = allocation.nextcall
-                allocation.nextcall = nextcall
-                allocation.already_accrued = False
-                if force_period and allocation.nextcall > date_to:
-                    allocation.nextcall = date_to
-                    force_period = False
-
-            # if plan.accrued_gain_time == 'start', process next period and set flag 'already_accrued', this will skip adding days
-            # once, preventing double allocation.
-            if allocation.accrual_plan_id.accrued_gain_time == 'start':
-                # check that we are at the start of a period, not on a carry-over or level transition date
-                level_starts = {level._get_level_transition_date(allocation.date_from): level for level in level_ids}
-                current_level = level_starts.get(allocation.actual_lastcall) or current_level or first_level
-                period_start = current_level._get_previous_date(allocation.actual_lastcall)
-                if allocation.actual_lastcall in {period_start, allocation.date_from} | set(level_starts.keys())\
-                        or (allocation.actual_lastcall - get_timedelta(current_level.accrual_validity_count, current_level.accrual_validity_type)
-                        in {period_start, allocation.date_from} | set(level_starts.keys())):
-                    period_end = current_level._get_next_date(allocation.lastcall)
-                    # Take level transition into account for the end_date
-                    end_date = None
-                    if allocation.accrual_plan_id.transition_mode == 'immediately':
-                        for level_idx, level in enumerate(level_ids):
-                            if level == current_level:
-                                if level_idx < len(level_ids) - 1:
-                                    end_date = min(period_end, level_ids[level_idx + 1]._get_level_transition_date(allocation.date_from))
-                                break
-                    if current_level.cap_accrued_time:
-                        current_level_maximum_leave = current_level.maximum_leave if current_level.added_value_type == "day"\
-                            else current_level.maximum_leave / allocation.employee_id._get_hours_per_day(allocation.lastcall)
-                    leaves_taken = _get_leaves_taken(allocation)
-                    ctx_allocation = allocation.with_context({'start_date': allocation.lastcall, 'end_date': end_date or period_end})
-                    ctx_allocation._add_days_to_allocation(current_level, current_level_maximum_leave, leaves_taken, period_start, period_end)
-                    allocation.already_accrued = True
+                if allocation._has_expiring_days(allocation_data) and allocation_data['carried_over_days_expiration_date'] > allocation_data['nextcall']:
+                    nextcall = min(allocation_data['carried_over_days_expiration_date'], nextcall)
 
-    @api.model
-    def _update_accrual(self):
-        """
-        Method called by the cron task in order to increment the number_of_days when
-        necessary.
+                if current_level.accrual_validity or allocation._has_expiring_days(allocation_data):
+                    # 3. Take expiring days into account
+                    if allocation_data['nextcall'] == allocation_data['carried_over_days_expiration_date']:
+                        allocation_data['allocated_duration'] = allocation._get_days_after_expiration(allocation_data)
+
+                if allocation.accrual_plan_id.accrued_gain_time == 'start':
+                    if allocation_data['nextcall'] == carryover_date:
+                        allocation_data['yearly_accrued_duration'] = 0
+                    allocation._process_accrual_start(allocation_data, accrual_level, next_accrual)
+                else:
+                    allocation._process_accrual_end(
+                        allocation_data, accrual_level, accrual_level_idx, first_level_start_date)
+                    if allocation_data['nextcall'] == carryover_date:
+                        allocation_data['yearly_accrued_duration'] = 0
+
+                allocation_data['lastcall'] = allocation_data['nextcall']
+                allocation_data['nextcall'] = nextcall
+
+            if allocation_data['leaves_taken'] is None:
+                allocation_data['leaves_taken'] = _get_leaves_taken(allocation, allocation_data)
+
+        for allocation in (accrual_allocations - accrual_plan_with_levels):
+            allocations_data[allocation]['leaves_taken'] = _get_leaves_taken(allocation, allocations_data[allocation])
+
+        return allocations_data
+
+    def _get_accrual_allocation_data(self):
+        allocation_data = {}
+        for allocation in self:
+            allocation_data[allocation] = {
+                'last_accrual': allocation.last_accrual,
+                'lastcall': allocation.lastcall,
+                'nextcall': allocation.nextcall,
+                'allocated_duration': allocation._convert_to_type_request_unit(allocation.number_of_days, 'day'),
+                'yearly_accrued_duration': allocation._convert_to_type_request_unit(allocation.yearly_accrued_days, 'day'),
+                'carried_over_days_expiration_date': allocation.carried_over_days_expiration_date,
+                'previous_carryover_allocated_duration':
+                    allocation._convert_to_type_request_unit(allocation.previous_carryover_number_of_days, 'day'),
+                # -------- Fields that won't be written on the allocation (see _update_accrual)
+                'leaves_taken': None,
+            }
+        return allocation_data
+
+    def _update_accrual(self, date_to=None):
+        """ Run by a cron in order to update the `number_of_days` (which represents the "credit" this allocation is granting)
+            depending on the linked accrual plan configuration
+            Update the accrual allocations until `date_to`
         """
-        today = datetime.combine(fields.Date.today(), time(0, 0, 0))
-        allocations = self.search([
+        allocations_data = self._process_accrual_plans(date_to)
+        to_update_fields = ('last_accrual', 'lastcall', 'nextcall', 'carried_over_days_expiration_date')
+        for allocation in self:
+            values = allocations_data[allocation]
+            to_update_values = {field: values[field] for field in to_update_fields}
+            to_update_values.update({
+                'yearly_accrued_days':
+                    allocation._convert_from_type_request_unit(values['yearly_accrued_duration'], 'day', values),
+                'previous_carryover_number_of_days':
+                    allocation._convert_from_type_request_unit(values['previous_carryover_allocated_duration'], 'day', values),
+            })
+            to_update_values['number_of_days'] = allocation._convert_from_type_request_unit(values['allocated_duration'], 'day', values)
+            allocation.write(to_update_values)
+
+    @api.model
+    def _get_to_update_accrual_allocations(self):
+        today = fields.Date.context_today(self)
+        return self.search([
             ('state', '=', 'validate'),
-            ('accrual_plan_id', '!=', False), ('employee_id', '!=', False),
-            '|', ('date_to', '=', False), ('date_to', '>', fields.Datetime.now()),
-            '|', ('nextcall', '=', False), ('nextcall', '<=', today)])
-        allocations._process_accrual_plans()
-
-    def _discard_fake_allocation(self):
-        """Discard temporary allocations used for accrual simulation.
-
-        Invalidating a new record clears its cache but does not remove it
-        from the pending recomputation queue. Since a new record has no
-        database row from which computed fields can be recomputed, remove
-        it from the recomputation queue before discarding it.
+            ('accrual_plan_id', '!=', False),
+            ('date_from', '<=', today),
+            '|', ('date_to', '=', False), ('date_to', '>=', today),
+            '|', ('nextcall', '=', False), ('nextcall', '<=', today),
+        ])
+
+    def _get_additionnal_future_leaves_on(self, accrual_date, allocations_data={}):
+        """ Possibly accrual inconsistent (see the comment under the class declaration)
+            :param allocations_data: dict containing the allocation mapped to their data. If `allocations_data` is a dict,
+            then this function will return the additionnal leaves compared to the leaves in the `allocations_data` (instead of
+            comparing it to `self.number_of_days` or `self.number_of_hours_display`)
+            :returns: added allocated duration expressed in the unit of the work entry type `unit_of_measure` of the allocation
         """
-        self.invalidate_recordset()
-        for field in list(self.env.transaction.tocompute):
-            if field.model_name == self._name:
-                self.env.remove_to_compute(field, self)
-
-    def _get_future_leaves_on(self, accrual_date):
-        # As computing future accrual allocation days automatically updates the allocation,
-        # We need to create a temporary copy of that allocation to return the difference in number of days
-        # to see how much more days will be allocated from now until that date.
         self.ensure_one()
         if not accrual_date or accrual_date <= date.today():
             return 0
 
+        if self in allocations_data:
+            allocation_data = allocations_data[self]
+            nextcall = allocation_data['nextcall']
+        else:
+            allocation_data = {}
+            nextcall = self.nextcall
         if not (self.accrual_plan_id
                 and self.state == 'validate'
                 and (not self.date_to or self.date_to > accrual_date)
-                and (not self.nextcall or self.nextcall <= accrual_date)):
+                and (not nextcall or nextcall <= accrual_date)):
             return 0
 
-        fake_allocation = self.env['hr.leave.allocation'].new(origin=self)
-        fake_allocation.sudo()._process_accrual_plans(accrual_date, log=False)
-        if self.work_entry_type_id.unit_of_measure == 'hour':
-            res = float_round(fake_allocation.number_of_hours_display - self.number_of_hours_display, precision_digits=2)
+        future_alloc_data = self.sudo()._process_accrual_plans(accrual_date, precomputed_allocations=allocations_data)[self]
+        if 'allocated_duration' in allocation_data:
+            current_allocated_duration = self._convert_from_type_request_unit(
+                allocation_data['allocated_duration'], self.work_entry_type_id.unit_of_measure, allocation_data)
         else:
-            res = round((fake_allocation.number_of_days - self.number_of_days), 2)
-        fake_allocation._discard_fake_allocation()
-        return res
+            current_allocated_duration = self._convert_duration(self.number_of_days, 'day', self.work_entry_type_id.unit_of_measure)
+        future_allocated_duration = self._convert_from_type_request_unit(
+            future_alloc_data['allocated_duration'], self.work_entry_type_id.unit_of_measure, allocation_data)
+        return float_round(future_allocated_duration - current_allocated_duration, precision_digits=2)
 
     def _get_next_states_by_state(self):
         self.ensure_one()
@@ -810,40 +861,6 @@ class HrLeaveAllocation(models.Model):
                 target=allocation.employee_id.name,
             )
 
-    def _add_lastcalls(self):
-        today = fields.Date.context_today(self)
-        for allocation in self:
-            if not allocation.accrual_plan_id or not allocation.accrual_plan_id.level_ids:
-                continue
-            (current_level, current_level_idx) = allocation._get_current_accrual_plan_level_id(today)
-            if not current_level:
-                first_level = allocation.accrual_plan_id.level_ids[0]
-                first_level_start = allocation.date_from + get_timedelta(first_level.start_count, first_level.start_type)
-                if first_level_start == today:
-                    current_level = first_level
-                    current_level_idx = 0
-            if not allocation.lastcall:
-                if not current_level:
-                    allocation.lastcall = today
-                    allocation.actual_lastcall = allocation.lastcall
-                    continue
-                allocation.lastcall = max(
-                    current_level._get_previous_date(today),
-                    allocation.date_from + get_timedelta(current_level.start_count, current_level.start_type)
-                )
-                allocation.actual_lastcall = allocation.lastcall
-            if current_level and not allocation.nextcall:
-                accrual_plan = allocation.accrual_plan_id
-                allocation.nextcall = current_level._get_next_date(allocation.lastcall)
-                if current_level_idx < (len(accrual_plan.level_ids) - 1) and accrual_plan.transition_mode == 'immediately':
-                    next_level = accrual_plan.level_ids[current_level_idx + 1]
-                    next_level_start = allocation.date_from + get_timedelta(next_level.start_count, next_level.start_type)
-                    allocation.nextcall = min(allocation.nextcall, next_level_start)
-                # If the expiration date didn't pass (expiration date is in the future)
-                expiration_date = allocation.carried_over_days_expiration_date
-                if expiration_date and expiration_date > allocation.lastcall:
-                    allocation.nextcall = min(allocation.nextcall, expiration_date)
-
     def add_follower(self, employee_id):
         employee = self.env['hr.employee'].browse(employee_id)
         if employee.user_id:
@@ -859,7 +876,6 @@ class HrLeaveAllocation(models.Model):
             if not values.get('department_id'):
                 values.update({'department_id': self.env['hr.employee'].sudo().browse(employee_id).department_id.id})
         allocations = super(HrLeaveAllocation, self.with_context(mail_create_nosubscribe=True)).create(vals_list)
-        allocations._add_lastcalls()
         for allocation in allocations:
             partners_to_subscribe = set()
             if allocation.employee_id.user_id:
@@ -885,10 +901,7 @@ class HrLeaveAllocation(models.Model):
         tracked_fields = {'number_of_days_display', 'number_of_hours_display', 'state', 'date_to'}
         tracked_updates = tracked_fields.intersection(values)
         if not tracked_updates:
-            res = super().write(values)
-            if 'accrual_plan_id' in values:
-                self._add_lastcalls()
-            return res
+            return super().write(values)
 
         def total_excess(extra_data):
             excess_days = extra_data.get('excess_days', {})
@@ -899,8 +912,6 @@ class HrLeaveAllocation(models.Model):
         result = super().write(values)
         consumed_leaves = self.employee_id._get_consumed_leaves(work_entry_types=self.work_entry_type_id)
 
-        if 'accrual_plan_id' in values:
-            self._add_lastcalls()
         for allocation in self:
             current_extra_data = dict(consumed_leaves[1]).get(allocation.employee_id, {}) \
                 .get(allocation.work_entry_type_id, {})
@@ -1030,16 +1041,14 @@ class HrLeaveAllocation(models.Model):
                 return False
         return True
 
-    def _get_initialize_accrual_plan_values(self, date_from):
+    def _get_init_accrual_plan_values(self):
         return {
-            'lastcall': date_from,
+            'last_accrual': False,
+            'lastcall': False,
             'nextcall': False,
             'number_of_days': 0.0,
-            'number_of_days_display': 0.0,
-            'number_of_hours_display': 0.0,
-            'already_accrued': False,
             'carried_over_days_expiration_date': False,
-            'expiring_carryover_days': 0
+            'previous_carryover_number_of_days': 0,
         }
 
     @api.onchange('accrual_plan_id')
@@ -1050,19 +1059,15 @@ class HrLeaveAllocation(models.Model):
             self.number_of_days = 1.0
 
     # Allows user to simulate how many days an accrual plan would give from a certain start date.
-    # it uses the actual computation function but resets values of lastcall, nextcall and nbr of days
-    # before every run, as if it was run from date_from, after an optional change in the allocation value
-    # the user can simply confirm and validate the allocation. The record is in correct state for the next
-    # call of the cron job.
     @api.onchange('date_from', 'accrual_plan_id', 'date_to', 'employee_id')
-    def _onchange_date_from(self):
+    def _onchange_process_accrual_plans(self):
         if not self.date_from or not self.accrual_plan_id or self.state == 'validate'\
            or not self.employee_id:
             return
-        update_vals = self._get_initialize_accrual_plan_values(self.date_from)
+        update_vals = self._get_init_accrual_plan_values()
         self.update(update_vals)
         date_to = min(self.date_to, date.today()) if self.date_to else False
-        self._process_accrual_plans(date_to)
+        self._update_accrual(date_to)
 
     # ------------------------------------------------------------
     # Activity methods

--- a/addons/hr_holidays/models/hr_work_entry_type.py
+++ b/addons/hr_holidays/models/hr_work_entry_type.py
@@ -7,8 +7,6 @@ import operator as py_operator
 from collections import defaultdict
 from datetime import date, datetime, time, UTC
 
-from dateutil.relativedelta import relativedelta
-
 from odoo import api, fields, models
 from odoo.exceptions import UserError, ValidationError
 from odoo.fields import Domain
@@ -642,32 +640,27 @@ been taken for this time off type. Changing it now would affect existing employe
 
     def _get_closest_expiring_leaves_date_and_count(self, allocations, remaining_leaves, target_date):
         # Get the expiration date and carryover date of all allocations and compute the closest expiration date
-        expiration_dates_per_allocation = defaultdict(lambda: {'expiration_date': fields.Date(), 'carryover_date': fields.Date(), 'carried_over_days_expiration_date': fields.Date()})
+        expiration_dates_per_allocation = {}
         expiration_dates = list()
         carried_over_days_expiration_data = self._get_carried_over_days_expiration_data(allocations, target_date)
         for allocation in allocations:
             expiration_date = allocation.date_to
 
-            accrual_plan_level = allocation.sudo()._get_current_accrual_plan_level_id(target_date)[0]
-            carryover_date = False
-            if accrual_plan_level and (accrual_plan_level.action_with_unused_accruals == 'lost'
-            or accrual_plan_level.carryover_options == 'limited'):
-                carryover_date = allocation.sudo()._get_carryover_date(target_date)
-                # If carry over date == target date, then add 1 year to carry over date.
-                # Rational: for example if carry over date = 01/01 this year and target date = 01/01 this year,
-                # then any accrued days on 01/01 this year will have their carry over date 01/01 next year
-                # and not 01/01 this year.
-                if carryover_date == target_date:
-                    carryover_date += relativedelta(years=1)
+            current_lvl, _ = allocation.sudo()._get_current_accrual_plan_level_idx(target_date)
+            carryover_date = None
+            if current_lvl and (current_lvl.action_with_unused_accruals == 'lost' or current_lvl.carryover_options == 'limited'):
+                carryover_date = allocation.sudo()._get_next_carryover_date(target_date, date_from_included=False)
 
             carried_over_days_expiration_date = carried_over_days_expiration_data[allocation]['expiration_date']
 
             expiration_dates.extend([expiration_date, carryover_date, carried_over_days_expiration_date])
-            expiration_dates_per_allocation[allocation]['expiration_date'] = expiration_date
-            expiration_dates_per_allocation[allocation]['carryover_date'] = carryover_date
-            expiration_dates_per_allocation[allocation]['carried_over_days_expiration_date'] = carried_over_days_expiration_date
+            expiration_dates_per_allocation[allocation] = {
+                'expiration_date': expiration_date,
+                'carryover_date': carryover_date,
+                'carried_over_days_expiration_date': carried_over_days_expiration_date,
+            }
 
-        expiration_dates = list(filter(lambda date: date is not False, expiration_dates))
+        expiration_dates = list(filter(bool, expiration_dates))
         expiration_dates.sort()
         # Compute the number of expiring leaves
         for closest_expiration_date in expiration_dates:
@@ -680,8 +673,10 @@ been taken for this time off type. Changing it now would affect existing employe
                 if expiration_date and expiration_date == closest_expiration_date:
                     expiring_leaves_count += remaining_leaves[allocation]['virtual_remaining_leaves']
                 elif carryover_date and carryover_date == closest_expiration_date:
-                    accrual_plan_level = allocation.sudo()._get_current_accrual_plan_level_id(target_date)[0]
-                    expiring_leaves_count += max(0, remaining_leaves[allocation]['virtual_remaining_leaves'] - accrual_plan_level.postpone_max_days)
+                    current_lvl = allocation.sudo()._get_current_accrual_plan_level_idx(target_date)[0]
+                    max_carried_over_duration = allocation._convert_from_type_request_unit(
+                        current_lvl.max_carriedover_duration, allocation.work_entry_type_id.unit_of_measure)
+                    expiring_leaves_count += max(0, remaining_leaves[allocation]['virtual_remaining_leaves'] - max_carried_over_duration)
                 elif carried_over_days_expiration_date and carried_over_days_expiration_date == closest_expiration_date:
                     expiring_leaves_count += carried_over_days_expiration_data[allocation]['no_expiring_days']
 
@@ -689,20 +684,25 @@ been taken for this time off type. Changing it now would affect existing employe
                 return closest_expiration_date, expiring_leaves_count
 
         # No leaves will expire
-        return False, 0
+        return None, 0
 
     def _get_carried_over_days_expiration_data(self, allocations, target_date):
-        fake_allocations = self.env['hr.leave.allocation']
+        updated_alloc_data = allocations.sudo()._process_accrual_plans(target_date)
+        carried_over_days_expiration_data = {}
         for allocation in allocations:
-            fake_allocations |= self.env['hr.leave.allocation'].new(origin=allocation)
-        fake_allocations.sudo()._process_accrual_plans(target_date, log=False)
-        carried_over_days_expiration_data = {
-            fake_allocation._origin:
-            {
-                'expiration_date': fake_allocation.carried_over_days_expiration_date,
-                'no_expiring_days': max(0, fake_allocation.expiring_carryover_days - fake_allocation.leaves_taken)
+            if not allocation.accrual_plan_id:
+                carried_over_days_expiration_data[allocation] = {
+                    'expiration_date': None,
+                    'no_expiring_days': 0,
+                }
+                continue
+            allocation_data = updated_alloc_data[allocation]
+            prev_carryover_allocated_duration = allocation._convert_from_type_request_unit(
+                allocation_data['previous_carryover_allocated_duration'], allocation.work_entry_type_id.unit_of_measure, allocation_data)
+            leaves_taken = allocation_data['leaves_taken']
+            carried_over_days_expiration_data[allocation] = {
+                'expiration_date': allocation_data['carried_over_days_expiration_date'],
+                'no_expiring_days':
+                    max(0, prev_carryover_allocated_duration - leaves_taken),
             }
-            for fake_allocation in fake_allocations
-        }
-        fake_allocations._discard_fake_allocation()
         return carried_over_days_expiration_data
